File: core/LongestIncreasingPathInAMatrix/LongestIncreasingPathInAMatrix.java
package com.dbc.code;

import java.util.Arrays;

public class LongestIncreasingPathInAMatrix {
    private int dfs(int row, int col, int[][] matrix, int[][] memo) {
        if (memo[row][col] != 0) return memo[row][col];
        memo[row][col]++;
        int[][] dirs = new int[][]{{row + 1, col}, {row - 1, col}, {row, col + 1}, {row, col - 1}};
        for (int[] dir : dirs) {
            if (dir[0] >= 0 && dir[0] < matrix.length && dir[1] >= 0 && dir[1] < matrix[0].length && matrix[row][col] < matrix[dir[0]][dir[1]]) {
                memo[row][col] = Math.max(memo[row][col], dfs(dir[0], dir[1], matrix, memo) + 1);
            }
        }
        return memo[row][col];
    }

    public int longestIncreasingPath(int[][] matrix) {
        int res = 0;
        int[][] memo = new int[matrix.length][matrix[0].length];
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                res = Math.max(res, dfs(i, j, matrix, memo));
            }
        }
        return res;
    }
}

File: core/LongestIncreasingPathInAMatrix/LongestIncreasingPathInAMatrix.py
from typing import List


def longestIncreasingPath(matrix: List[List[int]]) -> int:
    res, memo = 0, [[0] * len(matrix[0]) for _ in range(len(matrix))]
    rows, cols = len(matrix), len(matrix[0])
    def dfs(row, col):
        if memo[row][col] != 0:
            return memo[row][col]
        memo[row][col] += 1
        for new_row, new_col in [[row + 1, col], [row - 1, col], [row, col + 1], [row, col - 1]]:
            if 0 <= new_row < rows and 0 <= new_col < cols and matrix[new_row][new_col] > matrix[row][col]:
                memo[row][col] = max( memo[row][col], dfs(new_row, new_col) + 1)

        return memo[row][col]

    res = 0
    for i in range(rows):
        for j in range(cols):
            res = max(res, dfs(i, j))

    return res

